src/sdc11073/roles/waveformprovider/waveformproviderimpl.py

```diff
@@ -0,0 +1,293 @@
+
+        :param generator_class: the generator to be instantiated.
+        :param max_waveforms: limit number of waveforms.
+        :return: list of handles of created generators
+        """
+        name = self._mdib.data_model.pm_names.RealTimeSampleArrayMetricDescriptor
+        all_waveforms = self._mdib.descriptions.NODETYPE.get(name)
+        if max_waveforms:
+            all_waveforms = all_waveforms[:max_waveforms]
+        for waveform in all_waveforms:
+            min_value = 0
+            max_value = 1
+            try:
+                tech_range = waveform.TechnicalRange[0]
+            except IndexError:
+                pass
+            else:
+                if tech_range.Lower is not None:
+                    min_value = float(tech_range.Lower)
+                if tech_range.Upper is not None:
+                    max_value = float(tech_range.Upper)
+                if max_value < min_value:
+                    max_value, min_value = min_value, max_value  # swap values
+                elif min_value == max_value:
+                    max_value += 1
+            generator = generator_class(min_value=min_value,
+                                        max_value=max_value,
+                                        waveform_period=2.0,
+                                        sample_period=waveform.SamplePeriod)
+            self.register_waveform_generator(waveform.Handle, generator)
+        return [waveform.Handle for waveform in all_waveforms]
+
+    def _worker_thread_loop(self):
+        # delay start of operation
+        shall_stop = self._stop_worker.wait(timeout=self.notifications_interval)
+        if shall_stop:
+            return
+        timer = IntervalTimer(period_in_seconds=self.notifications_interval)
+        try:
+            while True:
+                shall_stop = self._stop_worker.wait(timeout=self.notifications_interval)
+                if shall_stop:
+                    return
+                behind_schedule_seconds = timer.wait_next_interval_begin()
+                self._log_waveform_timing(behind_schedule_seconds)
+                try:
+                    with self._mdib.rt_sample_transaction() as transaction:
+                        self.update_all_realtime_samples(transaction)
+                    self._log_waveform_timing(behind_schedule_seconds)
+                except Exception:  # noqa: BLE001
+                    # catch all to keep loop running
+                    self._logger.warning('could not update real time samples: %s', traceback.format_exc())
+        finally:
+            self._logger.info('rt_sample_sendloop end')
+
+    def _update_rt_samples(self, state: RealTimeSampleArrayMetricStateContainer):
+        """Update waveforms state from waveform generator (if available)."""
+        ctxt = Context(prec=10)
+        wf_generator = self._waveform_generators.get(state.DescriptorHandle)
+        if wf_generator:

```

Comment:
optional: this could also be written as `if wf_generator := self._waveform_generators.get(state.DescriptorHandle)`